src/intelligence_layer/connectors/retrievers/hybrid_qdrant_in_memory_retriever.py

```diff
@@ -0,0 +1,206 @@
+        self._distance_metric = distance_metric
+
+        self._search_client.set_sparse_model(sparse_model_name)
+        self._sparse_vector_field_name = "text-sparse"
+        self._dense_vector_field_name = "text-dense"
+
+        if self._search_client.collection_exists(collection_name=self._collection_name):
+            self._search_client.delete_collection(
+                collection_name=self._collection_name,
+            )
+
+        self._search_client.create_collection(
+            collection_name=self._collection_name,
+            vectors_config=VectorParams(size=128, distance=self._distance_metric),
+            sparse_vectors_config=self._search_client.get_fastembed_sparse_vector_params(),
+        )
+
+        self._add_texts_to_memory(documents)
+
+    def _sparse_embed_query(self, query: str) -> models.SparseVector:
+        if self._search_client.sparse_embedding_model_name is None:
+            raise ValueError("Sparse embedding model is not set!")
+        sparse_embedder = self._search_client._get_or_init_sparse_model(
+            model_name=self._search_client.sparse_embedding_model_name
+        )
+
+        sparse_vector = next(sparse_embedder.query_embed(query=query))
+        sparse_query_vector = models.SparseVector(
+            indices=sparse_vector.indices.tolist(),
+            values=sparse_vector.values.tolist(),
+        )
+
+        return sparse_query_vector
+
+    def get_relevant_documents_with_scores(
+        self, query: str, filter: Optional[models.Filter] = None
+    ) -> Sequence[SearchResult[int]]:
+        """Search for documents in memory using hybrid search.
+
+        This method first embeds the given query using both a dense and sparse embedding model.
+        Then, it performs two search requests, one for each embedding type.
+        The results are then fused using the Reciprocal Rank Fusion algorithm.
+
+        Args:
+            query: The text to be searched with.
+            filter: Optional conditions to filter by.
+
+        Returns:
+            All documents that correspond to the query and pass the filter,
+            sorted by their reciprocal rank fusion score.
+        """
+        dense_query_vector = self._embed(query, self._query_representation)
+        sparse_query_vector = self._sparse_embed_query(query)
+
+        dense_request = models.SearchRequest(
+            vector=models.NamedVector(
+                name=self._dense_vector_field_name,
+                vector=dense_query_vector,
+            ),
+            limit=self._k,
+            filter=filter,
+            with_payload=True,
+            score_threshold=self._threshold,
+        )
+        sparse_request = models.SearchRequest(
+            vector=models.NamedSparseVector(
+                name=self._sparse_vector_field_name,
+                vector=sparse_query_vector,
+            ),
+            limit=self._k,
+            filter=filter,
+            with_payload=True,
+        )
```

Comment:
self._threshold, according to the docs, is the minimal value of the fused score, but you don't apply any thresholding on the fused score and only use it with the dense request

Reply:
Unlike cosine-distance based retrieval scores, BM25 has unbouded scores, making score interpretations much more challenging. Hence, I decided to keep it simple and only apply this for the dense request. If you are aware of any approaches to solve these, please share them as I do believe this could be improved somehow.

Reply:
Ive talked to someone who has run experiments using bm + cosine in a qa document search setting. His experience is the following (paraphrased):
Threshold doesn't make sense, as while the cosine is theoretically 0...1, it is not trained to use the full range, so in reality its unclear what the realistic maximum values are. Therefore, setting a threshold has limited value. 

Therefore I'd like to either remove the threshold all together OR if you want to keep it, change the docs in the init so that it correctly describes what the threshold does.

He also mentioned that reciprocal rank fusion is fine, he had better success by simply normalizing the metrics to 1 on each request and then weighting all metrics equally. Thats something you may want to try.

Reply:
Just the insight I was looking for and it also matches my experience. I will default to removing it for simplicity, but not without experimenting with the suggested approach first, thanks!